beacon-chain/forkchoice/protoarray/optimistic_sync.go

```diff
@@ -0,0 +1,80 @@
+package protoarray
+
+import (
+	"fmt"
+
+	types "github.com/prysmaticlabs/eth2-types"
+	"github.com/prysmaticlabs/prysm/config/params"
+	"github.com/prysmaticlabs/prysm/encoding/bytesutil"
+)
+
+// This returns the minimum and maximum slot of the synced_tips tree
+func (f *ForkChoice) boundarySyncedTips() (types.Slot, types.Slot) {
+	f.syncedTips.RLock()
+	defer f.syncedTips.RUnlock()
+
+	min := params.BeaconConfig().FarFutureSlot
+	max := types.Slot(0)
+	for _, slot := range f.syncedTips.validatedTips {
+		if slot > max {
+			max = slot
+		}
+		if slot < min {
+			min = slot
+		}
+	}
+	return min, max
+}
+
+// Optimistic returns true if this node is optimistically synced
+// WARNING: this function does not check if slot corresponds to the
+//          block with the given root. The caller is responsible for
+//          checking this. A consensus bug could be a cause of getting
+//          this wrong, so think twice before passing a wrong pair.
+func (f *ForkChoice) Optimistic(root [32]byte, slot types.Slot) (bool, error) {
+
```

Comment:
Can you add a condition here? Error out if the slot is lower than the finalized epoch in store

Reply:
Actually maybe not. Thinking out loud, could you still get optimistic in that case?